src/rust/engine/fs/src/directory.rs

```diff
@@ -544,6 +552,114 @@ impl DigestTrie {
     }
   }
 
+  pub fn diff(&self, other: &DigestTrie) -> DigestTrieDiff {
+    let mut result = DigestTrieDiff {
+      our_unique_files: Vec::new(),
+      our_unique_dirs: Vec::new(),
+      their_unique_files: Vec::new(),
+      their_unique_dirs: Vec::new(),
+      changed_files: Vec::new(),
+    };
+
+    self.diff_helper(other, PathBuf::new(), &mut result);
+
+    result
+  }
+
+  fn diff_helper(&self, them: &DigestTrie, path_so_far: PathBuf, result: &mut DigestTrieDiff) {
+    let mut our_iter = self.0.iter();
+    let mut their_iter = them.0.iter();
+    let mut ours = our_iter.next();
+    let mut theirs = their_iter.next();
+
+    let add_unique =
+      |entry: &Entry, unique_files: &mut Vec<PathBuf>, unique_dirs: &mut Vec<PathBuf>| {
+        let path = path_so_far.join(entry.name().as_ref());
+        match entry {
+          Entry::File(_) => unique_files.push(path),
+          Entry::Directory(_) => unique_dirs.push(path),
+        }
+      };
+
```

Comment:
I tried making helpers for `add_ours` and `add_theirs`, but the helper would need mutable borrow to the relevant result vectors, which would block us borrowing them to pass to the recursive call.

I'm not sure how to get this working in Rust. If you have a suggestion I'm all ears. I don't like how repetitive the call sites are.

Reply:
In general, if a closure would borrow something for too long (because it "closes over" something mutable), it can be passed as an explicit argument, as you did here with the `unique_files` and `unique_dirs` arguments.

So I think that `add_ours` and `add_theirs` could take `&mut DigestTrieDiff` as an argument...?